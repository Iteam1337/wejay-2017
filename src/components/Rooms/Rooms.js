// @flow

import './Rooms.css'
import React, { Component } from 'react'
import { gql, graphql } from 'react-apollo'
import { withRouter } from 'react-router-dom'

export class Rooms extends Component {
  joinRoom = async name => {
    if (!localStorage.getItem('user')) {
      return
    }

    const { mutate } = this.props

    await mutate({
      variables: {
        input: { roomName: name, email: localStorage.getItem('user') }
      }
    })

    this.props.history.push(`/room/${name}`)
  }

  render () {
    const { rooms } = this.props

    if (!rooms.length) {
      return <div>No active rooms</div>
    }

    return (
      <ul className="Rooms">
        {rooms.map((room, i) => (
          <li className="Rooms__room" key={`room-${i}`}>
            <button
              className="Rooms__button"
              onClick={() => this.joinRoom(room.name)}
            >
              {room.name}
            </button>
          </li>
        ))}
      </ul>
    )
  }
}

const joinRoomMutation = gql`
  mutation joinRoom($input: JoinRoomInput!) {
    joinRoom(input: $input) {
      name
    }
  }
`

export default withRouter(graphql(joinRoomMutation)(Rooms))
