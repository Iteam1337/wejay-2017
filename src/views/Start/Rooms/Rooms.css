.Rooms__room:not(:last-child) {
  margin-bottom: 5px; }

.Rooms__button {
  appearance: none;
  background-color: #2d3339;
  border: 0;
  border-radius: 3px;
  color: #fff;
  cursor: pointer;
  font-size: 16px;
  padding: 10px 20px;
  transition: background-color 150ms ease-in-out, opacity 200ms ease-in-out;
  width: 100%; }
  .Rooms__button:hover {
    background-color: #363d45; }
  .Rooms__button:active, .Rooms__button:focus {
    background-color: #fcc169;
    outline: none; }
  .Rooms__button:disabled {
    cursor: not-allowed;
    opacity: 0.5; }
