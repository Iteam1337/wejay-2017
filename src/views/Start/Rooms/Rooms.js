// @flow

import * as WejayApi from '../__generated__/StartQuery'
import * as React from 'react'
import Button from '../../../components/Button/Button'
import styled from 'styled-components'

type Props = {
  joinRoom: (roomName: string) => Promise<void>,
  rooms: $PropertyType<WejayApi.StartQuery, "rooms">,
}

const RoomsList = styled.div`
  display: grid;
  grid-row-gap: 5px;
`

const RoomsTitle = styled.div`
  color: #333;
  margin-bottom: 20px;
`

const Rooms = ({ joinRoom, rooms }: Props) => {
  const hasRooms = rooms.length > 0

  return (
    <React.Fragment>
      <RoomsTitle>Rooms list</RoomsTitle>
      <RoomsList>
        {!hasRooms && <div>No active rooms</div>}

        {hasRooms &&
          rooms.map(room => (
            <Button key={room.name} onClick={() => joinRoom(room.name)}>
              {room.name}
            </Button>
          ))}
      </RoomsList>
    </React.Fragment>
  )
}

export default Rooms
