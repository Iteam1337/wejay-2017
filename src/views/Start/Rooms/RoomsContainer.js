// @flow

import * as storage from '../../../utils/storage'
import gql from 'graphql-tag'
import { graphql } from 'react-apollo'
import Rooms from './Rooms'
import { withRouter } from 'react-router-dom'

const JoinRoomMutation = gql`
  mutation JoinRoom($input: JoinRoomInput!) {
    joinRoom(input: $input) {
      name
    }
  }
`

const withJoinRoom = graphql(JoinRoomMutation, {
  props: ({ ownProps, mutate }) => ({
    joinRoom: async roomName => {
      const userId = localStorage.getItem('id')
      const { email } = await storage.getValue(userId)

      await mutate({
        variables: {
          input: { email, roomName },
        },
      })

      ownProps.history.push(`/room/${roomName}`)
    },
  }),
})

export default withRouter(withJoinRoom(Rooms))
