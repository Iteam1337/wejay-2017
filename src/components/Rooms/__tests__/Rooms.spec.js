import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Rooms } from '../Rooms'

describe('components/Rooms', () => {
  let component
  let instance
  let mutate
  let history

  beforeEach(() => {
    mutate = jest.fn()
    history = { push: jest.fn() }

    component = shallow(
      <Rooms history={history} mutate={mutate} rooms={[{ name: 'test' }]} />
    )

    instance = component.instance()
  })

  it('returns empty state if no rooms', () => {
    component.setProps({
      rooms: []
    })

    expect(toJson(component)).toMatchSnapshot()
  })

  it('renders Rooms', () => {
    expect(toJson(component)).toMatchSnapshot()
  })

  it('should handle button clicks', () => {
    const button = component.find('button')

    button.simulate('click')

    expect(mutate).toBeCalled()
  })

  describe('#joinRoom', () => {
    it('returns if no user is set', () => {
      localStorage.getItem.mockReturnValue(null)

      instance.joinRoom('test')

      expect(mutate).not.toBeCalled()
    })

    it('calls the add room mutation', () => {
      localStorage.getItem.mockReturnValue('cookie@monster.com')

      instance.joinRoom('test')

      expect(mutate).toHaveBeenCalledWith({
        variables: {
          input: {
            roomName: 'test',
            email: 'cookie@monster.com'
          }
        }
      })
    })

    it('redirects to the room', async () => {
      try {
        localStorage.getItem.mockReturnValue('cookie@monster.com')

        await instance.joinRoom('test')

        expect(history.push).toHaveBeenCalledWith('/room/test')
      } catch (e) {
        throw new Error(e)
      }
    })
  })
})
