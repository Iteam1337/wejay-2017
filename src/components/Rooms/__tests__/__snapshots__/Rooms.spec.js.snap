// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Rooms renders Rooms 1`] = `
<ul
  className="Rooms"
>
  <li
    className="Rooms__room"
  >
    <button
      className="Rooms__button"
      onClick={[Function]}
    >
      test
    </button>
  </li>
</ul>
`;

exports[`components/Rooms returns empty state if no rooms 1`] = `
<div>
  No active rooms
</div>
`;
